client/src/client.rs

```diff
@@ -215,14 +216,45 @@ impl Auth {
     }
 }
 
-pub trait RpcApi: Sized {
+pub struct Request<'c, T: 'static> {
+    method: &'static str,
+    args: Vec<serde_json::Value>, //TODO(stevenroose) something more efficient?
+    converter: &'static dyn Fn(&RawValue) -> Result<T>,
```

Comment:
I believe you can build this into the method signatures instead of explicitly calling a converter (e.g., https://github.com/lightningdevkit/rust-lightning/blob/5e86bbf97030c69b0118d1bef39554febfb1cbef/lightning-block-sync/src/rpc.rs#L40-L41).

Reply:
Yeah, the RPC example is assuming JSON responses, which could be examined and converted as needed (e.g. https://github.com/lightningdevkit/rust-lightning/blob/b89f8815c3d40d8c5bf9aa60af5b3fbb2ee78f25/lightning-block-sync/src/convert.rs#L42-L72).

But in the following REST example, the response could be either JSON or binary depending on the parameters, which I think is what you are alluding to.

https://github.com/lightningdevkit/rust-lightning/blob/b89f8815c3d40d8c5bf9aa60af5b3fbb2ee78f25/lightning-block-sync/src/rest.rs#L29-L35

There the response format needs to be explicit at the call site. For example:

https://github.com/lightningdevkit/rust-lightning/blob/b89f8815c3d40d8c5bf9aa60af5b3fbb2ee78f25/lightning-block-sync/src/rest.rs#L38-L58

So basically, you define conversion from bytes to some response format:

https://github.com/lightningdevkit/rust-lightning/blob/b89f8815c3d40d8c5bf9aa60af5b3fbb2ee78f25/lightning-block-sync/src/http.rs#L456-L478

And then any conversions from those formats to the desired types.